pkg/vf/virtio.go

```diff
@@ -289,6 +292,112 @@ func (dev *VirtioVsock) AddToVirtualMachineConfig(vmConfig *VirtualMachineConfig
+	if err := dev.validateNbdURI(dev.URI); err != nil {
+		return nil, fmt.Errorf("invalid NBD device 'uri': %s", err.Error())
+	}
+
+	if err := dev.validateNbdDeviceIdentifier(dev.DeviceIdentifier); err != nil {
+		return nil, fmt.Errorf("invalid NBD device 'deviceId': %s", err.Error())
+	}
+
+	attachment, err := vz.NewNetworkBlockDeviceStorageDeviceAttachment(dev.URI, dev.Timeout, dev.ReadOnly, dev.SyncronizationModeVZ())
+	if err != nil {
+		return nil, err
+	}
+
+	vzdev, err := vz.NewVirtioBlockDeviceConfiguration(attachment)
+	if err != nil {
+		return nil, err
+	}
+	err = vzdev.SetBlockDeviceIdentifier(dev.DeviceIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	return vzdev, nil
+}
+
+func (dev *NetworkBlockDevice) validateNbdURI(uri string) error {
+	if uri == "" {
+		return fmt.Errorf("'uri' must be specified")
+	}
+
+	parsed, err := url.Parse(uri)
+	if err != nil {
+		return fmt.Errorf("error: %w", err)
+	}
+
+	// The format specified by https://github.com/NetworkBlockDevice/nbd/blob/master/doc/uri.md
+	if parsed.Scheme != "nbd" && parsed.Scheme != "nbds" && parsed.Scheme != "nbd+unix" && parsed.Scheme != "nbds+unix" {
+		return fmt.Errorf("invalid scheme: %s. Expected one of: 'nbd', 'nbds', 'nbd+unix', or 'nbds+unix'", parsed.Scheme)
+	}
+
+	return nil
+}
+
+func (dev *NetworkBlockDevice) validateNbdDeviceIdentifier(deviceID string) error {
+	if deviceID == "" {
+		return fmt.Errorf("'deviceId' must be specified")
+	}
+
+	if strings.Contains(deviceID, "/") {
+		return fmt.Errorf("invalid 'deviceId': it cannot contain any forward slash")
+	}
+
+	if len(deviceID) > 255 {
+		return fmt.Errorf("invalid 'deviceId': exceeds maximum length")
+	}
+
+	return nil
+}
+
+func (dev *NetworkBlockDevice) SyncronizationModeVZ() vz.DiskSynchronizationMode {
+	if dev.SynchronizationMode == config.SynchronizationNoneMode {
+		return vz.DiskSynchronizationModeNone
+	}
+	return vz.DiskSynchronizationModeFull
+}
+
+func (dev *NetworkBlockDevice) AddToVirtualMachineConfig(vmConfig *VirtualMachineConfiguration) error {
+	storageDeviceConfig, err := dev.toVz()
+	if err != nil {
+		return err
+	}
+	log.Infof("Adding NBD device (uri: %s, deviceId: %s)", dev.URI, dev.DeviceIdentifier)
+	vmConfig.storageDevicesConfiguration = append(vmConfig.storageDevicesConfiguration, storageDeviceConfig)
+
+	return nil
+}
+
+func ListenNetworkBlockDevices(vm *VirtualMachine) error {
+	for _, dev := range vm.vfConfig.storageDevicesConfiguration {
+		attachment := dev.Attachment()
+		if nbdAttachment, isNbdAttachment := attachment.(*vz.NetworkBlockDeviceStorageDeviceAttachment); isNbdAttachment {
+			nbdID, err := dev.(*vz.VirtioBlockDeviceConfiguration).BlockDeviceIdentifier()
+			if err != nil {
+				log.Info("Found NBD device with no identifier. Please set a deviceId to each NBD device.")
+				return fmt.Errorf("NetworkBlockDevice must have deviceId. Error %v", err)
+			}
+			nbdConfig := vm.vfConfig.config.NetworkBlockDevice(nbdID)
+			if nbdConfig == nil {
```

Comment:
We go through quite a few steps (several interface -> concrete type conversions, one iteration over the VM devices to find the right ones, ...) because we want to print `nbdConfig.URI` in our log messages.

This `URI` is available directly from the `nbdAttachment` we get early, but the API for that is not implemented in Code-Hex/vz: https://developer.apple.com/documentation/virtualization/vznetworkblockdevicestoragedeviceattachment/4168501-url?language=objc
Maybe we could only print the nbd device ID in the logs for now, and file a bug or add the API to Code-Hex/vz, and then update this code? This would result in simpler code.

Reply:
An alternative with slightly simpler code is this:
```
diff --git a/pkg/vf/virtio.go b/pkg/vf/virtio.go
index 39af346..c2e78ef 100644
--- a/pkg/vf/virtio.go
+++ b/pkg/vf/virtio.go
@@ -29,6 +29,11 @@ type VirtioInput config.VirtioInput
 type VirtioGPU config.VirtioGPU
 type NetworkBlockDevice config.NetworkBlockDevice
 
+type vzNetworkBlockDevice struct {
+       *vz.VirtioBlockDeviceConfiguration
+       config *NetworkBlockDevice
+}
+
 func (dev *NVMExpressController) toVz() (vz.StorageDeviceConfiguration, error) {
        var storageConfig StorageConfig = StorageConfig(dev.StorageConfig)
        attachment, err := storageConfig.toVz()
@@ -315,7 +320,7 @@ func (dev *NetworkBlockDevice) toVz() (vz.StorageDeviceConfiguration, error) {
                return nil, err
        }
 
-       return vzdev, nil
+       return vzNetworkBlockDevice{VirtioBlockDeviceConfiguration: vzdev, config: dev}, nil
 }
 
 func (dev *NetworkBlockDevice) validateNbdURI(uri string) error {
@@ -372,17 +377,14 @@ func (dev *NetworkBlockDevice) AddToVirtualMachineConfig(vmConfig *VirtualMachin
 
 func ListenNetworkBlockDevices(vm *VirtualMachine) error {
        for _, dev := range vm.vfConfig.storageDevicesConfiguration {
-               attachment := dev.Attachment()
-               if nbdAttachment, isNbdAttachment := attachment.(*vz.NetworkBlockDeviceStorageDeviceAttachment); isNbdAttachment {
-                       nbdID, err := dev.(*vz.VirtioBlockDeviceConfiguration).BlockDeviceIdentifier()
-                       if err != nil {
-                               log.Info("Found NBD device with no identifier. Please set a deviceId to each NBD device.")
-                               return fmt.Errorf("NetworkBlockDevice must have deviceId. Error %v", err)
-                       }
-                       nbdConfig := vm.vfConfig.config.NetworkBlockDevice(nbdID)
-                       if nbdConfig == nil {
-                               return fmt.Errorf("no NBD device found with deviceId %s", nbdID)
+               if nbdDev, isNbdDev := dev.(vzNetworkBlockDevice); isNbdDev {
+                       nbdAttachment, isNbdAttachment := dev.Attachment().(*vz.NetworkBlockDeviceStorageDeviceAttachment)
+                       if !isNbdAttachment {
+                               log.Info("Found NBD device with no NBD attachment. Please file a vfkit bug.")
+                               return fmt.Errorf("NetworkBlockDevice must use a NBD attachment")
                        }
+
+                       nbdConfig := nbdDev.config
                        go func() {
                                for {
                                        select {
```

I've only compile-tested it, so I'm not 100% sure it works.

Reply:
Overall, your call, I'm fine with your initial code if you prefer to keep it this way.

Reply:
Applied your suggestion 👍